source/fe/mapping_q_cache.cc

```diff
@@ -125,6 +184,319 @@ MappingQCache<dim, spacedim>::initialize(
+
+      typename DoFHandler<dim, spacedim>::cell_iterator cell(
+        &cell_tria->get_triangulation(),
+        cell_tria->level(),
+        cell_tria->index(),
+        &dof_handler);
+
+      const auto mapping_q_generic =
+        dynamic_cast<const MappingQGeneric<dim, spacedim> *>(&mapping);
+
+      // Step 2a) set up FEValues (if needed)
+      if (((vector_describes_relative_displacement ||
+            (is_relevant_cell == false)) &&
+           ((mapping_q_generic != nullptr &&
+             this->get_degree() == mapping_q_generic->get_degree()) ==
+            false)) ||
+          ((is_relevant_cell == true) &&
+           (((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree()) ==
+            false)))
+        {
+          // get FEValues (thread-safe); in the case that this thread has
+          // not created a an FEValues object yet, this helper-function also
+          // creates one with the right quadrature rule
+          auto &fe_values = fe_values_all.get();
+          if (fe_values.get() == nullptr)
+            {
+              QGaussLobatto<dim> quadrature_gl(this->polynomial_degree + 1);
+
+              std::vector<Point<dim>> quadrature_points;
+              for (const auto i :
+                   FETools::hierarchic_to_lexicographic_numbering<dim>(
+                     this->polynomial_degree))
+                quadrature_points.push_back(quadrature_gl.point(i));
+              Quadrature<dim> quadrature(quadrature_points);
+
+              const auto &fe_temp =
+                ((is_relevant_cell == true) &&
+                 (((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree()) ==
+                  false)) ?
+                  fe :
+                  static_cast<const FiniteElement<dim, spacedim> &>(fe_nothing);
+
+              fe_values = std::make_unique<FEValues<dim, spacedim>>(
+                mapping, fe_temp, quadrature, update_quadrature_points);
+            }
+
+          fe_values->reinit(cell);
+        }
+
+      std::vector<Point<spacedim>> result;
+
+      // Step 2b) read of quadrature points in the relative displacement case
+      if (vector_describes_relative_displacement || (is_relevant_cell == false))
+        {
+          if (mapping_q_generic != nullptr &&
+              this->get_degree() == mapping_q_generic->get_degree())
+            result =
+              mapping_q_generic->compute_mapping_support_points(cell_tria);
+          else
+            result = fe_values_all.get()->get_quadrature_points();
+
+          if (is_relevant_cell == false)
+            return result;
+        }
+      else
+        {
+          result.resize(
+            Utilities::pow<unsigned int>(this->get_degree() + 1, dim));
+        }
+
+      // Step 2c) read global vector and adjust points accordingly
+      if ((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree())
+        {
+          AssertDimension(dim, spacedim); // TODO: codim-1
```

Comment:
Can someone give me a hint how to approach the problem of codim meshes here (`dim!=spacedim` - my application of this function). At this stage and later on in the general case where we will call `FEValues::get_function_values()`, we will have points of size dim and of size spacedim. Obviously, we need to multiply it with some Jacobian, but where do I get this one? Any hint (maybe in the form of a reference to other classes) are welcome!

Reply:
Actually, I think this should be not a problem since the `FESystem` is set up with dim components.